### ISSUE TITLE
Comparing CUDA with Pytorch

### ISSUE THREAD

**Issue by NamGyuKang** [NamGyuKang/CosineSampler#1]
![Helmholtz_2d_release](https://user-images.githubusercontent.com/70684829/206615560-77b358ff-59d7-4ef8-af99-48312514c1ab.png)

**Comment by YeongHyeon** [comment]
Hi, NamGyu

I think that the PyTorch framework includes CUDA acceleration.
What is the difference between CUDA and PyTorch here?

Thank you for your kindness.

**Comment by NamGyuKang** [comment]
Pytorch's grid_sample() has CUDA accelerated code. However, this method only supports first-order backpropagation. This means that we cannot handle or use interpolated tensors that are differentiated in the model.

In our PIXEL research, we needed a third-order gradient of the interpolated tensor. An important contribution is that third-order backpropagation is possible. In other words, it works even if the loss function is composed of the second derivative of the interpolated tensor. I created a cosine_sampler_backward_backward_backward_kernel( ) function in the CUDA code.

Also, the kernel type required cosine interpolation in addition to linear interpolation. Furthermore In the PIXEL research, adding different small offsets value to the grid (the second input component of grid_sample()) was needed. 

As a result, we can enjoy the triple backward sampler by choosing what we want with the argument options (kernel = {bilinear, trilinear, cosine, smooth-step}, multicell = {True, False})

I sincerely thank you very much for your interest in the PIXEL study and the CosineSampler study.

If you have any more questions, please feel free to ask ^^

Sincerely,
Namgyu Kang

**Comment by YeongHyeon** [comment]
Wow!
This study is really interesting.
Thank you for your response.

Best regards,
YeongHyeon Park

**Comment by NamGyuKang** [comment]
Thank you again for your interest!

Sincerely,
Namgyu Kang
